### PR TITLE
Implemented Spring tests

### PR DESCRIPTION
**Pull request by Stelmaschyk**

### CODE REVIEW

File: src/test/java/mate/academy/bookstore/utill/BookProvider.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+import java.math.BigDecimal;
+import java.util.Set;
+import java.util.stream.Collectors;
+import mate.academy.bookstore.dto.book.CreateBookRequestDto;
+import mate.academy.bookstore.dto.book.UpdateBookRequestDto;
+import mate.academy.bookstore.model.Book;
+import mate.academy.bookstore.model.Category;
+
+public class BookProvider {
```
Comment (okuzan) [inline src/test/java/mate/academy/bookstore/utill/BookProvider.java]: you can also prevent instantiation of util class

File: src/test/java/mate/academy/bookstore/services/CategoryServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        when(bookMapper.toBookDto(books.get(1))).thenReturn(expectedBookDto.get(1));
+
+        List<BookDto> actual = categoryService.getBooksByCategoriesId(TEST_CATEGORY_ID);
+        assertNotNull(actual);
+        assertEquals(expectedBookDto.size(), actual.size());
+
+        verify(categoryRepository, times(1)).existsById(TEST_CATEGORY_ID);
+        verify(bookRepository, times(1)).findAllByCategoriesId(TEST_CATEGORY_ID);
+        verify(bookMapper, times(1)).toBookDto(books.get(0));
+        verify(bookMapper, times(1)).toBookDto(books.get(1));
```
Comment (okuzan) [inline src/test/java/mate/academy/bookstore/services/CategoryServiceTest.java]: times(1) is default, we can omit this

File: src/test/java/mate/academy/bookstore/services/CategoryServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+                .updateCategoryById(TEST_CATEGORY_ID, updateCategoryRequestDto);
+
+        assertEquals(expected.getId(), updatedCategoryResponseDto.getId());
+        assertEquals(expected.getName(), updatedCategoryResponseDto.getName());
+
+        verify(categoryMapper, times(1))
+                .updateCategoryFromDto(updateCategoryRequestDto, category);
+        verify(categoryRepository, times(1)).save(category);
+        verify(categoryMapper, times(1))
+                .updateCategoryFromDto(updateCategoryRequestDto, category);
```
Comment (okuzan) [inline src/test/java/mate/academy/bookstore/services/CategoryServiceTest.java]: repeated twice

File: src/test/java/mate/academy/bookstore/services/CategoryServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+            """)
+    @Test
+    void findAllCategories_WithValidRequestDto_ShouldReturnCategoriesResponseDto() {
+        Pageable pageable = PageRequest.of(0, 10);
+        Page<Category> categoryPage = new PageImpl<Category>(categories, pageable,
+                categories.size());
+
+        when(categoryRepository.findAll(pageable)).thenReturn(categoryPage);
+
+        List<CategoryResponseDto> expectedBookDtos = categories.stream()
```
Comment (okuzan) [inline src/test/java/mate/academy/bookstore/services/CategoryServiceTest.java]: misleading name, book vs category

File: src/test/java/mate/academy/bookstore/services/CategoryServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        when(categoryRepository.findById(TEST_INCORRECT_CATEGORY_ID)).thenReturn(Optional.empty());
+
+        EntityNotFoundException entityNotFoundException =
+                assertThrows(EntityNotFoundException.class,
+                    () -> categoryService.getById(TEST_INCORRECT_CATEGORY_ID));
+
+        String expectedMessage = "Can't find category by id " + TEST_INCORRECT_CATEGORY_ID;
+        String actualMessage = entityNotFoundException.getMessage();
+
+        assertEquals(expectedMessage, actualMessage);
```
Comment (okuzan) [inline src/test/java/mate/academy/bookstore/services/CategoryServiceTest.java]:
```suggestion
        String actualMessage = entityNotFoundException.getMessage();

        assertTrue(actualMessage.contains(String.valueOf(TEST_INCORRECT_CATEGORY_ID)));
```
we can make it less dependent on specific wording while still being useful

File: src/test/java/mate/academy/bookstore/services/BookServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        verify(bookMapper, times(1)).toBookDto(book);
+        verifyNoMoreInteractions(bookMapper, bookRepository);
+    }
+
+    @Test
+    @DisplayName("""
+        Find all books with a valid request
+        Should return a list of BookDto
+            """)
+    public void findAllBooks_WithValidCreateBookRequestDto_ShouldReturnListOfBookDto() {
```
Comment (okuzan) [inline src/test/java/mate/academy/bookstore/services/BookServiceTest.java]: fix name, method doesn't use CreateBookRequestDto

File: src/test/java/mate/academy/bookstore/controller/BookControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        BookDto actual = objectMapper.readValue(
+                result.getResponse().getContentAsString(), BookDto.class);
+        EqualsBuilder.reflectionEquals(expected, actual);
+    }
+
+    @WithMockUser(username = "admin", roles = {"ADMIN"})
+    @Test
+    @DisplayName("delete book by id")
+    void deleteById_WithValidBookId_Success() throws Exception {
+        MvcResult result = mockMvc.perform(delete("/books/{id}", 4)
```
Comment (okuzan) [inline src/test/java/mate/academy/bookstore/controller/BookControllerTest.java]:
```suggestion
        mockMvc.perform(delete("/books/{id}", 4)
```

File: src/test/java/mate/academy/bookstore/controller/CategoryControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        Assertions.assertEquals(TEST_CATEGORY_ID, actual.getId());
+    }
+
+    @WithMockUser(username = "user")
+    @Test
+    @DisplayName("Get books list by category id")
+    @Sql(scripts = "classpath:database/categories_books/add-books-to-categories.sql",
+            executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+    @Sql(scripts = "classpath:database/categories_books"
+                + "/delete-data-from-categories-books-table.sql",
```
Comment (okuzan) [inline src/test/java/mate/academy/bookstore/controller/CategoryControllerTest.java]:
```suggestion
    @Sql(scripts = "classpath:database/categories_books/clear-categories-books-table.sql",
```
let's try to fit it in one line

File: src/test/java/mate/academy/bookstore/controller/CategoryControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        JsonNode rootNode = objectMapper.readTree(result.getResponse().getContentAsString());
+        JsonNode contentNode = rootNode.get("content");
+        CategoryResponseDto[] actual = objectMapper.treeToValue(contentNode,
+            CategoryResponseDto[].class);
+
+        Assertions.assertEquals(TEST_CATEGORY_AMOUNT, actual.length);
```
Comment (okuzan) [inline src/test/java/mate/academy/bookstore/controller/CategoryControllerTest.java]: looks a bit complex, will this work?
```suggestion
Page<CategoryResponseDto> page = objectMapper.readValue(
    result.getResponse().getContentAsString(),
    new TypeReference<Page<CategoryResponseDto>>() {}
);
Assertions.assertEquals(TEST_CATEGORY_AMOUNT, page.getContent().size());
```

File: pom.xml [hunk 10]
```diff
@@ -119,7 +120,46 @@
...
+            <artifactId>spring-test</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.security</groupId>
+            <artifactId>spring-security-test</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+
```
Comment (okuzan) [inline pom.xml]:
```suggestion
```

**Review by okuzan — CHANGES_REQUESTED** [review]
Good job, some comments

File: src/test/java/mate/academy/bookstore/services/CategoryServiceTest.java (hunk 5 above)
Comment (Stelmaschyk) [reply src/test/java/mate/academy/bookstore/services/CategoryServiceTest.java]: nice idea

File: src/test/java/mate/academy/bookstore/controller/BookControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+        UpdateBookRequestDto updateDto = BookProvider.updateRequestDto(TEST_ID);
+        Book expected = BookProvider.updateBook(updateDto);
+        MvcResult result = mockMvc.perform(put("/books/{id}", TEST_ID)
+                .content(objectMapper.writeValueAsString(updateDto))
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+        BookDto actual = objectMapper.readValue(
+                result.getResponse().getContentAsString(), BookDto.class);
+        EqualsBuilder.reflectionEquals(expected, actual);
```
Comment (sarakhmen) [inline src/test/java/mate/academy/bookstore/controller/BookControllerTest.java]: I don't see any assertions here

File: src/test/java/mate/academy/bookstore/controller/BookControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+                .andReturn();
+
+        BookDto[] actual = objectMapper.readValue(
+            result.getResponse().getContentAsString(), BookDto[].class);
+        List<Long> bookIds = Arrays.stream(actual)
+                .map(BookDto::getId)
+                .toList();
+
+        Assertions.assertEquals(EXPECTED_LENGTH, actual.length);
+        assertThat(bookIds).containsExactlyInAnyOrder(1L, 2L, 3L, 4L);
```
Comment (sarakhmen) [inline src/test/java/mate/academy/bookstore/controller/BookControllerTest.java]: we can make these 2 assertions in one chain. AssertJ is a powerful library and it provides us such a possibility

File: src/test/java/mate/academy/bookstore/controller/BookControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+    void getById_WithValidBookId_ShouldReturnBookDto() throws Exception {
+        MvcResult result = mockMvc.perform(get("/books/{id}", TEST_ID)
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        BookDto actual = objectMapper.readValue(
+                result.getResponse().getContentAsString(), BookDto.class);
+        Assertions.assertNotNull(actual);
+        Assertions.assertEquals(TEST_ID, actual.getId());
```
Comment (sarakhmen) [inline src/test/java/mate/academy/bookstore/controller/BookControllerTest.java]: let's use the same assertion library across all tests. I'd recommend using asserJ as previously. Fix in all places

**Review by sarakhmen — CHANGES_REQUESTED** [review]
let's improve your solution :)

File: src/test/java/mate/academy/bookstore/controller/BookControllerTest.java (hunk 12 above)
Comment (Stelmaschyk) [reply src/test/java/mate/academy/bookstore/controller/BookControllerTest.java]: not bad library

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**